### PR TITLE
use express mounting, kill connect-cookie-session

### PR DESCRIPTION
**Pull request by seanmonstar**
I was looking through wsapi, and the cascade of functions made go all like `huh?`. We can mount middleware at routes, removing the need for url checking of `/wsapi` in each middleware, and simple mount it at the correct prefix.

This makes it make more sense (imo).

For good measure, ripped out connect-cookie-session, and just used client-sessions.

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
GitHub is lying, all the tests are green.

File: lib/wsapi.js [hunk 1]
```diff
@@ -461,10 +441,11 @@ exports.setup = function(options, app) {
...
 
-  app.use(checkCodeVersion);
+  app.use(WSAPI_PREFIX, checkCodeVersion);
 
-  app.use(function(req, resp, next) {
+  app.use(WSAPI_PREFIX, function(req, resp, next) {
     var operation = operationFromURL(req.url);
+    
     // not a WSAPI request
     if (!operation) return next();
```
Comment (shane-tomlinson) [inline lib/wsapi.js:450]: Should we log if an operation is not found to see if somebody is trying to find a back door?

File: scripts/serve_example_primary.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ path = require('path'),
 urlparse = require('urlparse'),
 postprocess = require('postprocess'),
 querystring = require('querystring'),
-sessions = require('connect-cookie-session'),
```
Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:14]: +10000

File: scripts/serve_example_primary.js [hunk 3]
```diff
@@ -19,25 +19,20 @@ require("jwcrypto/lib/algs/rs");
...
-        secure: false,
-        maxAge: 1 * 60 * 60 * 1000
-      }
-    })(req, res, next);
-  } else {
-    next();
+exampleServer.use(API_PREFIX, sessions({
+  secret: "this secret, isn't very secret",
+  cookieName: 'example_browserid_primary',
+  cookie: {
```
Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:32]: Should a `duration` be configured as well as `max-age`? See https://github.com/mozilla/browserid/pull/3980/files#diff-43bc087e8ce30c866ff58c56df77be13R352

File: scripts/serve_example_primary.js [hunk 4]
```diff
@@ -69,12 +60,12 @@ exampleServer.get("/api/whoami", function (req, res) {
 });
 
 exampleServer.get("/api/login", function (req, res) {
-  req.session = {user: req.query.user};
+  req.session.user = req.query.user;
```
Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:66]: @seanmonstar, this causes a problem when testing manually:

```
example_primary (44079): TypeError: Cannot set property 'user' of undefined
example_primary (44079):     at exampleServer.post.user (/Users/stomlinson/development/browserid/scripts/serve_example_primary.js:63:20)
example_primary (44079):     at callbacks (/Users/stomlinson/development/browserid/node_modules/express/lib/router/index.js:272:11)
example_primary (44079):     at param (/Users/stomlinson/development/browserid/node_modules/express/lib/router/index.js:246:11)
example_primary (44079):     at pass (/Users/stomlinson/development/browserid/node_modules/express/lib/router/index.js:253:5)
example_primary (44079):     at Router._dispatch (/Users/stomlinson/development/browserid/node_modules/express/lib/router/index.js:280:5)
example_primary (44079):     at Object.Router.middleware [as handle] (/Users/stomlinson/development/browserid/node_modules/express/lib/router/index.js:45:10)
example_primary (44079):     at next (/Users/stomlinson/development/browserid/node_modules/connect/lib/http.js:201:15)
example_primary (44079):     at Object.exampleServer.get.req.session.user [as handle] (/Users/stomlinson/development/browserid/scripts/serve_example_primary.js:54:3)
example_primary (44079):     at next (/Users/stomlinson/development/browserid/node_modules/connect/lib/http.js:201:15)
example_primary (44079):     at Object.bodyParser [as handle] (/Users/stomlinson/development/browserid/node_modules/connect/lib/middleware/bodyParser.js:59:61)
```

It looks like session is not yet set, so we have to do this the original way.

When I revert this change, I then receive an error in provisioning the provisioning iframe when /api/whoami is called, the response is `null`

Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:66]: More info regarding the call to `/api/whoami`:

From the Firefox dev console:

```
Request URL:    http://127.0.0.1:10005/api/whoami
Request Method:     GET
Status Code:    HTTP/1.1 200 OK
Request Headers 16:56:14.000
X-Requested-With:   XMLHttpRequest
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:27.0) Gecko/20100101 Firefox/27.0
Referer:    http://127.0.0.1:10005/provision.html
Host:   127.0.0.1:10005
DNT:    1
Connection: keep-alive
Accept-Language:    en,de;q=0.8,es;q=0.6,en-gb;q=0.4,chrome://global/locale/intl.properties;q=0.2
Accept-Encoding:    gzip, deflate
Accept: */*
Sent Cookie
can_set_cookies:    1
Response Headers Δ1ms
X-Powered-By:   Express
Date:   Wed, 16 Oct 2013 15:56:14 GMT
Content-Type:   application/json; charset=utf-8
Content-Length: 4
Connection: keep-alive
Cache-Control:  no-store, max-age=0
Response Body Δ0ms
null
```

It looks like the only cookie that was sent was `can_set_cookies`.

Going back a bit further to the call to `/api/login`, it looks like no cookies are sent:

```

Request URL:    http://127.0.0.1:10005/api/login?user=shane
Request Method:     GET
Status Code:    HTTP/1.1 200 OK
Request Headers 16:56:12.000
X-Requested-With:   XMLHttpRequest
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:27.0) Gecko/20100101 Firefox/27.0
Referer:    http://127.0.0.1:10005/sign_in.html?email=shane%40example.domain
Host:   127.0.0.1:10005
DNT:    1
Connection: keep-alive
Accept-Language:    en,de;q=0.8,es;q=0.6,en-gb;q=0.4,chrome://global/locale/intl.properties;q=0.2
Accept-Encoding:    gzip, deflate
Accept: */*
Sent Cookie
can_set_cookies:    1
Response Headers Δ1ms
X-Powered-By:   Express
Date:   Wed, 16 Oct 2013 15:56:12 GMT
Content-Type:   application/json; charset=utf-8
Content-Length: 4
Connection: keep-alive
Cache-Control:  no-store, max-age=0
Response Body Δ0ms
null
```

Comment (seanmonstar) [inline scripts/serve_example_primary.js:66]: Oh, I think I forgot to include requestKey to client sessions. So it used
the cookie name.

Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:66]: ah, I don't see anything about requestKey in the documentation for https://github.com/mozilla/node-client-sessions, we should add it.

Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:66]: With this diff, everything works as expected:

``` diff
diff --git a/scripts/serve_example_primary.js b/scripts/serve_example_primary.js
index 51711bd..79debe9 100755
--- a/scripts/serve_example_primary.js
+++ b/scripts/serve_example_primary.js
@@ -20,17 +20,20 @@ require("jwcrypto/lib/algs/ds");

 var exampleServer = express.createServer();
 const API_PREFIX = '/api';
+const SESSION_DURATION_MS = 1 * 60 * 60 * 1000;

 exampleServer.use(express.cookieParser());

 exampleServer.use(API_PREFIX, sessions({
   secret: "this secret, isn't very secret",
   cookieName: 'example_browserid_primary',
+  requestKey: 'session',
+  duration: SESSION_DURATION_MS,
   cookie: {
     path: '/api',
     httpOnly: true,
     secure: false,
-    maxAge: 1 * 60 * 60 * 1000
+    maxAge: SESSION_DURATION_MS
   }
 }));

```

Comment (shane-tomlinson) [inline scripts/serve_example_primary.js:66]: documented requestKey https://github.com/mozilla/node-client-sessions/pull/60
